### DISCUSSION TITLE
How to load local built Pyodide in Node.js

### DISCUSSION THREAD

**Discussion by LijieZhang1998** [pyodide/pyodide#4071] [Q&A] [answered]
Hi, I'm new to Pyodide. I followed the instruction to build pyodide from source. I noticed that `node_modules` and `dist/pyodide.js` file are created. But how does Node.js use the local built pyodide? I didn't see `pyodide` module under `node_modules` folder? So when I ran the below code, I got an error `Error: Cannot find module 'pyodide'`. If I ran `npm install` to install `pyodide` first, the code is working. I didn't find any doc talking about it. Any help is appreciated. 
```
const test = function () {
  const { loadPyodide } = require("pyodide");
  async function hello_python() {
    let pyodide = await loadPyodide();
    await pyodide.loadPackage("numpy");
    const src = `import numpy as np

x = np.arange(15, dtype=np.int64).reshape(3, 5)
x[1:, ::2] = -99
print(x)
# array([[  0,   1,   2,   3,   4],
#        [-99,   6, -99,   8, -99],
#        [-99,  11, -99,  13, -99]])

print(x.max(axis=1))
# array([ 4,  8, 13])

# Generate normally distributed random numbers:
rng = np.random.default_rng()
samples = rng.normal(size=10)
print(samples)`;
    return pyodide.runPythonAsync(src);
  }
  hello_python().then((result) => {
    console.log("Python says that 1+1 =", result);
  });
};

test();
```

**Comment by hoodmane** [comment]
Good question. For example if you have the following folder layout:
```
pyodide/
   dist/
   src/
   ... rest of source tree
myproject/
   node_modules/
```
Then from inside of `myproject` use `npm i ../pyodide/dist`. On a linux, this is actually the same as `ln -s ../pyodide/dist node_modules/pyodide`. The `node_modules` folder inside the Pyodide source tree is for building Pyodide not for running it, ideally you should leave that one alone.

**Reply by LijieZhang1998** [reply]
Thank you. I followed your folder structure. `require("pyodide")` is working. But `await pyodide.loadPackage("numpy")` throws `Error: No known package with name 'numpy'`. The installed `node_modules` in `myproject` folder doesn't have `numpy` module. It's in `pyodide/packages` folder. How do I use the packages in `pyodide/packages` folder?

**Reply by hoodmane** [reply, marked as answer]
You need to build it first. E.g. first use, `PYODIDE_PACKAGES="tag:core, numpy" make`. After that loading numpy will work.

**Reply by LijieZhang1998** [reply]
Yea, it's working now. I built `numpy` on my local docker container, and was able to load it. Thank you very much. I feel there is not much doc content talking about pyodide usage with Node.js. If you want me to add some doc about it, I'll be happy for it.

**Reply by hoodmane** [reply]
we very much appreciate any contributions to the docs!
